File: src/2024/01.ts
// https://adventofcode.com/2024/day/1

function cutLocations(input: string): [number[], number[]] {
  const aa = input.split('\n').map(line => line.split(/\s+/).map(Number))
  return [aa.flatMap(a => a[0]), aa.flatMap(a => a[1])]
}

export function partOne(input: string): number {
  const [one, two] = cutLocations(input).map(a => a.toSorted())
  let sum = 0
  for (let i = 0; i < one.length; i++)
    sum += Math.abs(one[i] - two[i])

  return sum
}

export function partTwo(input: string): number {
  const [one, two] = cutLocations(input).map(a => a.toSorted())

  const maxes: Record<number, number> = {}
  for (const num of two) {
    if (maxes[num] === undefined)
      maxes[num] = 1
    else maxes[num]++
  }

  let sum = 0
  for (const num of one)
    sum += num * (maxes[num] || 0)

  return sum
}

File: src/2024/01.kt
import java.io.File
import kotlin.math.abs

fun main(args : Array<String>) {
    val fileContent = File("src/2024/01.input").readText()
    val lines = fileContent.split("\n")
    val parsedLines = lines.map { it.split("\\s+".toRegex()).map { it.toInt() } }
    val one = parsedLines.map { it[0] }.sorted()
    val two = parsedLines.map { it[1] }.sorted()

    var sumPartOne = 0
    for (i in one.indices)
        sumPartOne += abs(one[i] - two[i])
    println(sumPartOne)

    val maxes = mutableMapOf<Int,Int>()
    for (num in two) {
        val cur: Int = maxes.get(num) ?: 0
        maxes.put(num, cur + 1)
    }
    var sumPartTwo = 0
    for (num in one)
        sumPartTwo += num * (maxes[num] ?: 0)

    println(sumPartTwo)
}
